src/Microsoft.ML.ImageAnalytics/ImageLoader.cs

```diff
@@ -201,11 +242,107 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
                                 throw Host.Except($"Failed to load image {src.ToString()}.");
                         }
                     };
+
+                return del;
+            }
+
+            private Delegate MakeGetterVectorDataViewByteType(DataViewRow input, int iinfo, Func<int, bool> activeOutput, out Action disposer)
+            {
+                Contracts.AssertValue(input);
+                Contracts.Assert(0 <= iinfo && iinfo < _parent.ColumnPairs.Length);
+
+                disposer = null;
+                var getSrc = input.GetGetter<ReadOnlyMemory<char>>(input.Schema[ColMapNewToOld[iinfo]]);
+                ReadOnlyMemory<char> src = default;
+                ValueGetter<VBuffer<byte>> del =
+                    (ref VBuffer<byte> dst) =>
+                    {
+                        byte[] buffer = null;
+                        bool addBufferBack = false;
+                        if (!_bufferPool.TryTake(out buffer))
+                        {
+                            buffer = new byte[4096];
+                            addBufferBack = true;
+                        }
+
+                        getSrc(ref src);
+
+                        if (src.Length > 0)
+                        {
+                            string path = src.ToString();
+                            if (!string.IsNullOrWhiteSpace(_parent.ImageFolder))
+                                path = Path.Combine(_parent.ImageFolder, path);
+                            if (!TryLoadDataIntoBuffer(path, ref dst, buffer))
+                                throw Host.Except($"Failed to load image {src.ToString()}.");
+                        }
+
+                        if (addBufferBack)
+                        {
+                            Contract.Assert(buffer != null);
+                            _bufferPool.Add(buffer);
+                        }
+                    };
+
                 return del;
             }
 
+            private static bool TryLoadDataIntoBuffer(string path, ref VBuffer<byte> imgData, byte[] readBuffer)
```

Comment:
Here's how I'd expect this code to look:

https://github.com/eerhardt/machinelearning/commit/779a1ddef737456dcbb9dc5121b4f8fde501501d.

Have you tried that approach?

Reply:
Yes I have tried it. I see failures for ArrayPool in the Azure machines. I may have missed changing Dependencies.prop though. The performance gain was similar for current approach and ArrayPool. Around 58 secs for both approaches.

Reply:
Can you use the above code then? If you are getting failures in CI, we can fix those. 
My proposed code is simpler, plus it doesn’t use an intermediate buffer on the latest .NET versions.

Your current approach will throw every other buffer away since you don’t return the buffer to the pool if you took it from the pool.

Reply:
Eric is right, we are not putting the buffer back into concurrent bag, I recommend passing concurrent bag to this function so that after it is done it puts the array back in the bag. I'm suggesting this since you are getting CI failures with ArrayPool.

---
In reply to: [330830331](https://github.com/dotnet/machinelearning/pull/4242#discussion_r330830331) [](ancestors = 330830331)

Reply:
I would suggest not to use the ConcurrentBag at all. Can you show me the errors with the ArrayPool approach?